root/programs/C#/Frameworks/Infrastructure/Business/Presentation/MyBaseApiController.cs

```diff
@@ -0,0 +1,361 @@
+        /// <summary>エラーログの出力</summary>
+        /// <param name="filterContext">ExceptionContext</param>
+        private void OutputErrorLog(ExceptionContext filterContext)
+        {
+            // 非同期ControllerのInnerException対策（底のExceptionを取得する）。
+            Exception ex = filterContext.Exception;
+            Exception bottomException = ex;
+            while (bottomException.InnerException != null)
+            {
+                bottomException = bottomException.InnerException;
+            }
+
+            // ------------
+            // メッセージ部
+            // ------------
+            // ユーザ名, IPアドレス,
+            // レイヤ, 画面名, コントロール名, 処理名
+            // 処理時間（実行時間）, 処理時間（CPU時間）
+            // エラーメッセージID, エラーメッセージ等
+            // ------------
+
+            string strLogMessage =
+                "," + (this.UserInfo != null ? this.UserInfo.UserName : "null") +
+                "," + (this.UserInfo != null ? this.UserInfo.IPAddress : "null") +
+                "," + "----->>" +
+                "," + this.ControllerName +
+                "," + this.ActionName + "(OnException)" +
+                "," + //this.perfRec.ExecTime +
+                "," + //this.perfRec.CpuTime + 
+                "," + GetExceptionMessageID(bottomException) +
+                "," + bottomException.Message + "\r\n"+
+                "," + bottomException.StackTrace + "\r\n" +
+                "," + ex.ToString(); // Exception.ToString()はRootのExceptionに対して行なう。
+
+            LogIF.ErrorLog("ACCESS", strLogMessage);
+        }
+
+        /// <summary>ExceptionのMessageIDを返す。</summary>
+        /// <param name="ex">Exception</param>
+        /// <returns>ExceptionのMessageID</returns>
+        private string GetExceptionMessageID(Exception ex)
+        {
+            // Check exception type 
+            if (ex is BusinessSystemException)
+            {
+                // システム例外
+                BusinessSystemException bsEx = (BusinessSystemException)ex;
+                return bsEx.messageID;
+            }
+            else if (ex is FrameworkException)
+            {
+                // フレームワーク例外
+                FrameworkException fxEx = (FrameworkException)ex;
+                return fxEx.messageID;
+            }
+            else
+            {
+                // それ以外の例外
+                return "other Exception";
+            }
+        }
+
+        #endregion
+
+        #region 情報取得用
+
+        /// <summary>ユーザ情報を取得する</summary>
+        private void GetUserInfo(ActionExecutingContext filterContext)
+        {
+            // nullチェック
+            if (this.UserInfo == null)
+            {
+                // nullの場合、仮の値を生成 / 設定する。
+                string userName = System.Threading.Thread.CurrentPrincipal.Identity.Name;
+
+                if (userName == null || userName == "")
+                {
+                    // 未認証状態
+                    this.UserInfo = new MyUserInfo("未認証", filterContext.RequestContext.HttpContext.Request.UserHostAddress);
```

Comment:
@daisukenishino 

`Request.UserHostAddress` returns IPv6 address.
Isn't it need to obtain IPv4?

Reply:
@daisukenishino 

[This site](http://gooner.hateblo.jp/entry/2015/10/05/clientip/) says as follows:

```txt
OWIN を使っている場合は、MS_HttpContext ではなく、MS_OwinContext の RemoteIpAddress から取得します。
```

The case that uses owin is considered?